### DISCUSSION TITLE
Video "PyTorch Fundamentals" - Having a difference of output

### DISCUSSION THREAD

**Discussion by Arcane-WD** [mrdbourke/pytorch-deep-learning#990] [Q&A]
Hello, I have a small doubt in the vid part "PyTorch and NumPy". I create a new ones tensor, then create a variable to hold numpy version of said tensor, then increment tensor using `tensor = tensor + 1` there is no data shared b/w tensor and numpy_tensor. But if I use `tensor += 1`, the data is shared and both show the same output.
![image_2024-06-28_115621661](https://github.com/mrdbourke/pytorch-deep-learning/assets/119485294/73149687-380d-4190-ae9f-2852fba69967)

**Comment by hillions** [comment]
From the video, it looks like `array = array + 1 ` creates a new `numpy.ndarray`, while doing something like `array[0] = 9` will not create a new one. (keeping the memory shared) You can check this by looking at the memory location of the array:

```Python
array = np.arange(1.0, 8.0)
print(array)
print(f"NumPy array location: {hex(id(array))}")
array[0] = 9
print(array)
print(f"NumPy array location: {hex(id(array))}")
array = array + 1
print(array)
print(f"NumPy array location: {hex(id(array))}")
```

Gives an output like this:
```
[1. 2. 3. 4. 5. 6. 7.]
NumPy array location: 0x7fd468e05bd0
[9. 2. 3. 4. 5. 6. 7.]
NumPy array location: 0x7fd468e05bd0
[10.  3.  4.  5.  6.  7.  8.]
NumPy array location: 0x7fd468e11810
```

I'm guessing that when you create a `numpy.ndarray`, then a `torch.Tensor` from it using `torch.from_numpy()`, then the `torch.Tensor` still references the old `array` memory even after you create a new one with `array = array + 1`. They have become disassociated when you do `array = array + 1`.

**Comment by udaykumarswamy** [comment]
Hey @Arcane-WD 

here what i found from the documentation of `torch.numpy()`

torch.Tensor.numpy
Tensor.numpy(*, force=False) → [numpy.ndarray](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray)
Returns the tensor as a NumPy ndarray.

If force is False (the default), the conversion is performed only if the tensor is on the CPU, does not require grad, does not have its conjugate bit set, and is a dtype and layout that NumPy supports. **The returned ndarray and the tensor will share their storage, so changes to the tensor will be reflected in the ndarray and vice versa.**

i hope that should resolve your doubt.
